### ISSUE TITLE
Don't you think that using `Result` it's better to use specific incarnation of `Error` instead of just `Error`

### ISSUE THREAD

**Issue by valeriyvan** [NSHipster/articles#611]
It's about [Optional, throws, Result, and async/await](https://nshipster.com/optional-throws-result-async-await/).

Don't you think that using `Result` it's better to use specific incarnation of `Error` instead of just `Error`?

I think this `func keychainData(service: String) -> Result<Data, KeychainError >`
is better that this
`func keychainData(service: String) -> Result<Data, Error>'.

Signature clearly communicates what kind of error is in `Result`.

cc: @jemmons

**Comment by jemmons** [comment]
It's totally reasonable for us to be specific with our `Result`'s failure type as long as two things are true:

1. Only one type of error occurs in our function.
1. We only ever need to compose our `Result` with other `Result`s that share the same failure type.

Note that, while it's possible for us to be completely confident about requirement 1,* we can't really make any grantees about requirement 2 unless we mark it `final private` and are the only people working in the code base.

For example, in the blog post, all of `keychainData`'s errors are of the `KeychainError` type. But later we `flatMap` the `keychainData` `Result` with `makeAvatar` and `save`, neither of which return `Result`s with failure types of `KeychainError`. If we want them to be able to work together, they all have to share the lowest common denominator of failure type, which is the `Error` protocol.

This is a [good programming practice](https://tuhrig.de/programming-to-an-interface/) in general. Favoring the abstract `Error` interface over the specific `KeychainErrror` implementation not only keeps our code loosely coupled, but also mirrors how Swift treats errors when using the built-in error handling. We don't qualify the type a function throws for example:

```
// We can't tell Swift "this only throws KeychainError":
func keychainData(service: String) throws KeychainError -> Data { ... } 🛑

// Instead, we only say "this throws" and it's assumed to throw an `Error`:
func keychainData(service: String) throws -> Data { ... } ✅
```

In theory, we might think this means we're tossing away a ton of information by downcasting our specific errors to general ones. But in practice, when we handle errors, we're always doing it in the context of a `switch` or `catch`, and we're usually matching errors by type anyway — which has the effect of upcasting the error back to its original type:

```
enum FeatureError: Error {
  case fooError
  
  func handleError() {
    // ...
  }
}

func foo() throws {
  throw FeatureError.fooError
}

do {
  try foo()
} catch let error as FeatureError {
  //Totally works!
  error.handleError()
} catch {
  print("something else went wrong")
}
```

So to sum up: circumstances in which we could use specific error types in practice are more rare than we might think. But that's ok because it doesn't feel like we lose a lot by keeping our failure type general, and we actually get a little more flexibility and composability as a bonus.

---
\* Though, as an aside, I think this is a bit of a contrived example. Most of the time, in the wild, my functions contain a variety of error types. If for no other reason than they're passing through errors from different dependencies while also adding their own errors to the mix.

**Comment by mattt** [comment]
I agree with @jemmons', and think that's a great articulation of the trade-off of composibility over specificity in the declaration. 

See also [this thread from the Swift mailing list / forum](https://forums.swift.org/t/proposal-typed-throws/64/5) (among others) in which John McCall (who designed and implemented error handling in Swift) offers compelling argument against typed `throws`. 

Personally, I remember being disappointed that `throws` didn't have an explicit type when I first saw it. I thought it'd be really nice for `try / catch` to offer the same exhaustiveness checks that the compiler provides for `switch` statements. But ultimately, I found John's rationale to be convincing, and have come to agree with this design decision.

**Comment by jemmons** [comment]
The phrase “aspirational bluff“ as applied to language design is 💯.
